pkg/db/db_test.go

```diff
@@ -0,0 +1,274 @@
+package db
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"testing"
+	"time"
+
+	"golang.org/x/xerrors"
+	"k8s.io/utils/clock"
+	clocktesting "k8s.io/utils/clock/testing"
+
+	"github.com/aquasecurity/trivy-db/pkg/db"
+	"github.com/aquasecurity/trivy/pkg/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+type MockConfig struct {
+	mock.Mock
+}
+
+func (_m *MockConfig) GetMetadata() (db.Metadata, error) {
+	ret := _m.Called()
+	ret0 := ret.Get(0)
+	if ret0 == nil {
+		return db.Metadata{}, ret.Error(1)
+	}
+	metadata, ok := ret0.(db.Metadata)
+	if !ok {
+		return db.Metadata{}, ret.Error(1)
+	}
+	return metadata, ret.Error(1)
+}
+
+type MockGitHubClient struct {
+	mock.Mock
+}
+
+func (_m *MockGitHubClient) DownloadDB(ctx context.Context, fileName string) (io.ReadCloser, error) {
+	ret := _m.Called(ctx, fileName)
+	ret0 := ret.Get(0)
+	if ret0 == nil {
+		return nil, ret.Error(1)
+	}
+	rc, ok := ret0.(io.ReadCloser)
+	if !ok {
+		return nil, ret.Error(1)
+	}
+	return rc, ret.Error(1)
+}
+
+func TestClient_Download(t *testing.T) {
+	type getMetadataOutput struct {
+		metadata db.Metadata
+		err      error
+	}
+
+	type downloadDBOutput struct {
+		fileName string
+		err      error
+	}
+	type downloadDB struct {
+		input  string
+		output downloadDBOutput
+	}
+
+	testCases := []struct {
+		name            string
+		light           bool
+		clock           clock.Clock
+		getMetadata     []getMetadataOutput
```

Comment:
I'm curious as to why is `getMetadata` a slice of `getMetadataOutput`? Can there ever be more than one behaviour at a time for `getMetadata` in a single invocation? It also looks like all the test cases here only have a single element anyway.

Reply:
I assumed a case where `getMetadata` was not called. In that case, I need to avoid calling `Mock.On().Return()`. This is because we use `AssertExpectations`. If the slice is empty, we can skip the `for` loop. But for now, `getMetadata` is always called once. We can change from a slice to an object.

Reply:
Fixed.